# Treating Missing Values

There are broadly two ways to treat missing values:
1. Delete: Delete the missing values
2. Impute:
    * Imputing by a simple statistic: Replace the missing values by another value, commonly the mean, median, mode etc.
    * Predictive techniques: Use statistical models such as k-NN, SVM etc. to predict and impute missing values.
   
In general, imputaion makes assumptions about the missing values and replaces missing values by arbitrary numbers such as mean, median etc. It should be used only when you are reasonably confident about the assumptions.

Otherwise , deletion is often safer and recommended. You may lost some data, but will not make any unreasonable assumptions.

In [82]:
import numpy as np
import pandas as pd

df = pd.read_csv('datasets/melbourne.csv')
df

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St      2    h        NaN     SS   
1        Abbotsford        85 Turner St      2    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St      2    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St      3    u        NaN     VB   
4        Abbotsford        5 Charles St      3    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23542  Wyndham Vale     25 Clitheroe Dr      3    u        NaN     PN   
23543  Wyndham Vale    19 Dalrymple Bvd      4    h        NaN      S   
23544     Yallambie        17 Amaroo Wy      4    h  1100000.0      S   
23545    Yarraville          6 Agnes St      4    h  1285000.0     SP   
23546    Yarraville       33 Freeman St      4    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  ...  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0  ...       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0  ...       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0  ...       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0  ...       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0  ...       2.0  0.0   
...              ...         ...       ...       ...  ...       ...  ...   
23542      Harcourts  26-08-2017      27.2    3024.0  ...       1.0  0.0   
23543  hockingstuart  26-08-2017      27.2    3024.0  ...       NaN  NaN   
23544     Buckingham  26-08-2017      12.7    3085.0  ...       3.0  2.0   
23545        Village  26-08-2017       6.3    3013.0  ...       1.0  1.0   
23546        Village  26-08-2017       6.3    3013.0  ...       2.0  2.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea Lattitude  Longtitude  \
0         126.0           NaN        NaN        Yarra -37.80140   144.99580   
1         202.0           NaN        NaN        Yarra -37.79960   144.99840   
2         156.0          79.0     1900.0        Yarra -37.80790   144.99340   
3           0.0           NaN        NaN        Yarra -37.81140   145.01160   
4         134.0         150.0     1900.0        Yarra -37.80930   144.99440   
...         ...           ...        ...          ...       ...         ...   
23542     552.0         119.0     1990.0          NaN -37.90032   144.61839   
23543       NaN           NaN        NaN          NaN -37.87882   144.60184   
23544       NaN           NaN        NaN          NaN -37.72006   145.10547   
23545     362.0         112.0     1920.0          NaN -37.81188   144.88449   
23546       NaN         139.0     1950.0          NaN -37.81829   144.87404   

                  Regionname Propertycount  
0      Northern Metropolitan        4019.0  
1      Northern Metropolitan        4019.0  
2      Northern Metropolitan        4019.0  
3      Northern Metropolitan        4019.0  
4      Northern Metropolitan        4019.0  
...                      ...           ...  
23542   Western Metropolitan        5262.0  
23543   Western Metropolitan        5262.0  
23544  Northern Metropolitan        1369.0  
23545   Western Metropolitan        6543.0  
23546   Western Metropolitan        6543.0  

[23547 rows x 21 columns]

# Treating Missing Values in Columns

In [83]:
# summing up the missing values (column-wise)
round(100 * (df.isnull().sum() / len(df.index)), 2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
BuildingArea     57.46
YearBuilt        50.99
CouncilArea      33.51
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64

Notice that there are columns having almost 22%, 19%, 26%, 57% etc missing values. When dealing with columns, you have two simple choices - either **delete or retain the data(column)**. If you retain the column, you'll have to treat the rows having missing values.

If you delete the missing rows, you lost the data. If you impute, you introduce bias.

Apart from the number of missing values, the decision to delete or reatin a variable depends on various other factors, such as:

* The analysis task at hand
* The usefulness of the variable (based on your understanding of the problem)
* The total size of available data(if you have enough, you can afford to throw away some of it) etc

For e.g. Let's say that we want to build a (linear regression) model to predict the house prices in Melbourne. Now, even though the variable `Price` has about 22% missing values, you cannot drop the variable, since that is what we want to predict.

Simlarly, you would expect some other variables such as `Bedroom2`, `Bathroom`, `Landsize` etc. to be important predictors of `Price` and thus cannot be removed.

There are others such as `Building Area` which although seem important, have more than 50% missing values. It is impossible to either delete or impute the rows corresponding to such large number of missing values without losing a lot of data or introducing heavy bias.

Thus, for this exercise, let's remove columns having more than 30% missing values, i.e. `BuildingArea`, `YearBuilt`, `CouncilArea`.


In [84]:
df = df.drop('BuildingArea', axis=1)
df = df.drop('YearBuilt', axis = 1)
df = df.drop('CouncilArea', axis = 1)


round(100 * (df.isnull().sum()/len(df.index)), 2)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64

# Treat missing Values in Rows

In [85]:
df.isnull().sum(axis=1)

0        1
1        0
2        0
3        1
4        0
        ..
23542    1
23543    5
23544    1
23545    0
23546    1
Length: 23547, dtype: int64

In [86]:
df

Suburb             Address  Rooms Type      Price Method  \
0        Abbotsford       68 Studley St      2    h        NaN     SS   
1        Abbotsford        85 Turner St      2    h  1480000.0      S   
2        Abbotsford     25 Bloomburg St      2    h  1035000.0      S   
3        Abbotsford  18/659 Victoria St      3    u        NaN     VB   
4        Abbotsford        5 Charles St      3    h  1465000.0     SP   
...             ...                 ...    ...  ...        ...    ...   
23542  Wyndham Vale     25 Clitheroe Dr      3    u        NaN     PN   
23543  Wyndham Vale    19 Dalrymple Bvd      4    h        NaN      S   
23544     Yallambie        17 Amaroo Wy      4    h  1100000.0      S   
23545    Yarraville          6 Agnes St      4    h  1285000.0     SP   
23546    Yarraville       33 Freeman St      4    h  1050000.0     VB   

             SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
0             Jellis  03-09-2016       2.5    3067.0       2.0       1.0  1.0   
1             Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2             Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
3             Rounds  04-02-2016       2.5    3067.0       3.0       2.0  1.0   
4             Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
...              ...         ...       ...       ...       ...       ...  ...   
23542      Harcourts  26-08-2017      27.2    3024.0       3.0       1.0  0.0   
23543  hockingstuart  26-08-2017      27.2    3024.0       NaN       NaN  NaN   
23544     Buckingham  26-08-2017      12.7    3085.0       4.0       3.0  2.0   
23545        Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   
23546        Village  26-08-2017       6.3    3013.0       4.0       2.0  2.0   

       Landsize  Lattitude  Longtitude             Regionname  Propertycount  
0         126.0  -37.80140   144.99580  Northern Metropolitan         4019.0  
1         202.0  -37.79960   144.99840  Northern Metropolitan         4019.0  
2         156.0  -37.80790   144.99340  Northern Metropolitan         4019.0  
3           0.0  -37.81140   145.01160  Northern Metropolitan         4019.0  
4         134.0  -37.80930   144.99440  Northern Metropolitan         4019.0  
...         ...        ...         ...                    ...            ...  
23542     552.0  -37.90032   144.61839   Western Metropolitan         5262.0  
23543       NaN  -37.87882   144.60184   Western Metropolitan         5262.0  
23544       NaN  -37.72006   145.10547  Northern Metropolitan         1369.0  
23545     362.0  -37.81188   144.88449   Western Metropolitan         6543.0  
23546       NaN  -37.81829   144.87404   Western Metropolitan         6543.0  

[23547 rows x 18 columns]

Notice that now, we have removed most of the rows where multiple columns(`Bedroom2`, `Bathroom`, `Landsize`) were missing.
Now, we still have about 21% missing values in the column `Price` and 9% in `Landsize`. Since `Price` still contains a lot of missing data (and imputing 21% values of a variable you want to predict will introduce heavy bias), its a bad idea to impute those values.

Thus, let's remove the missing rows in `Price` as well. Notice that you can use `np.isnan(df['column'])` to filter out the corresponding rows, and use a ~ to discard the values satisfying the condintion.

In [87]:
df.shape

(23547, 18)

In [88]:
# removing NaN Price rows
df = df[~np.isnan(df['Price'])]

round(100 * (df.isnull().sum() / len(df.index)), 2)


Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price             0.00
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.01
Postcode          0.01
Bedroom2         18.86
Bathroom         18.87
Car              19.44
Landsize         26.05
Lattitude        18.11
Longtitude       18.11
Regionname        0.01
Propertycount     0.01
dtype: float64

In [89]:
df.shape

(18396, 18)

In [90]:
len(df)

18396

Now, we have Landsize as the only variable having a significant number of missing values. Let's give this variable a chance and consider imputing the NaNs.


The decision (whether and how to impute) will depend upon the distribution of the variable. For e.g., if the variable is such that all the observations lie in a short range (say between 800 sq.ft to 820 sq.ft), you can take a call to impute the missing values by something like that mean or median Landsize.

Let's look at the distribution

In [91]:
df['Landsize'].describe()

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
25%         176.500000
50%         440.000000
75%         651.000000
max      433014.000000
Name: Landsize, dtype: float64

Notice that the minumum is 0, max is 433104, the mean is 558 and median(50%) is 440. There's a significant variation in the 25th and the 75th percentile as well(176, 651).

Thus, imputing this with mean/median seems quite biased, and so we should remove the NaNs.

In [92]:
df = df[~np.isnan(df['Landsize'])]

round(100 * (df.isnull().sum() / len(df.index)), 2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.46
Landsize         0.00
Lattitude        0.16
Longtitude       0.16
Regionname       0.00
Propertycount    0.00
dtype: float64

In [95]:
df.loc[:, ['Lattitude', 'Longtitude']].describe()

Lattitude    Longtitude
count  13581.000000  13581.000000
mean     -37.809204    144.995221
std        0.079257      0.103913
min      -38.182550    144.431810
25%      -37.856820    144.929600
50%      -37.802360    145.000100
75%      -37.756400    145.058320
max      -37.408530    145.526350

In [96]:
df['Lattitude'].fillna(df['Lattitude'].mean(), inplace=True)

In [97]:
df

Suburb           Address  Rooms Type      Price Method  \
1         Abbotsford      85 Turner St      2    h  1480000.0      S   
2         Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
4         Abbotsford      5 Charles St      3    h  1465000.0     SP   
5         Abbotsford  40 Federation La      3    h   850000.0     PI   
6         Abbotsford       55a Park St      4    h  1600000.0     VB   
...              ...               ...    ...  ...        ...    ...   
23537  Wheelers Hill      12 Strada Cr      4    h  1245000.0      S   
23538   Williamstown     77 Merrett Dr      3    h  1031000.0     SP   
23539   Williamstown       83 Power St      3    h  1170000.0      S   
23541   Williamstown      96 Verdon St      4    h  2500000.0     PI   
23545     Yarraville        6 Agnes St      4    h  1285000.0     SP   

        SellerG        Date  Distance  Postcode  Bedroom2  Bathroom  Car  \
1        Biggin  03-12-2016       2.5    3067.0       2.0       1.0  1.0   
2        Biggin  04-02-2016       2.5    3067.0       2.0       1.0  0.0   
4        Biggin  04-03-2017       2.5    3067.0       3.0       2.0  0.0   
5        Biggin  04-03-2017       2.5    3067.0       3.0       2.0  1.0   
6        Nelson  04-06-2016       2.5    3067.0       3.0       1.0  2.0   
...         ...         ...       ...       ...       ...       ...  ...   
23537     Barry  26-08-2017      16.7    3150.0       4.0       2.0  2.0   
23538  Williams  26-08-2017       6.8    3016.0       3.0       2.0  2.0   
23539     Raine  26-08-2017       6.8    3016.0       3.0       2.0  4.0   
23541   Sweeney  26-08-2017       6.8    3016.0       4.0       1.0  5.0   
23545   Village  26-08-2017       6.3    3013.0       4.0       1.0  1.0   

       Landsize  Lattitude  Longtitude                  Regionname  \
1         202.0  -37.79960   144.99840       Northern Metropolitan   
2         156.0  -37.80790   144.99340       Northern Metropolitan   
4         134.0  -37.80930   144.99440       Northern Metropolitan   
5          94.0  -37.79690   144.99690       Northern Metropolitan   
6         120.0  -37.80720   144.99410       Northern Metropolitan   
...         ...        ...         ...                         ...   
23537     652.0  -37.90562   145.16761  South-Eastern Metropolitan   
23538     333.0  -37.85927   144.87904        Western Metropolitan   
23539     436.0  -37.85274   144.88738        Western Metropolitan   
23541     866.0  -37.85908   144.89299        Western Metropolitan   
23545     362.0  -37.81188   144.88449        Western Metropolitan   

       Propertycount  
1             4019.0  
2             4019.0  
4             4019.0  
5             4019.0  
6             4019.0  
...              ...  
23537         7392.0  
23538         6380.0  
23539         6380.0  
23541         6380.0  
23545         6543.0  

[13603 rows x 18 columns]

In [98]:
df['Lattitude'].isnull().any()

False

In [101]:
df['Longtitude'].fillna(df['Longtitude'].mean(), inplace=True)

In [102]:
df['Longtitude'].isnull().any()

False

In [105]:
df.loc[:, ['Bathroom', 'Car']].describe()

Bathroom           Car
count  13602.000000  13540.000000
mean       1.534921      1.610414
std        0.691834      0.962244
min        0.000000      0.000000
25%        1.000000      1.000000
50%        1.000000      2.000000
75%        2.000000      2.000000
max        8.000000     10.000000

In [106]:
df['Longtitude'].mean()

144.99522086076092

These two are integer type variables, and thus have values 0, 1, 2 etc. We cannot impute the NaNs by the mean or median (1.53 bathrooms does not make sense!).

Thus, we need to impute them by the mode - the most common occuring value.

In [109]:
# converting to type 'category'
df['Car'] = df['Car'].astype('category')

df['Car'].head()

1    1.0
2    0.0
4    0.0
5    1.0
6    2.0
Name: Car, dtype: category
Categories (11, float64): [0.0, 1.0, 2.0, 3.0, ..., 7.0, 8.0, 9.0, 10.0]

In [110]:
df['Bathroom'] = df['Bathroom'].astype('category')

df['Bathroom'].head()

1    1.0
2    1.0
4    2.0
5    2.0
6    1.0
Name: Bathroom, dtype: category
Categories (9, float64): [0.0, 1.0, 2.0, 3.0, ..., 5.0, 6.0, 7.0, 8.0]

In [111]:
df.dtypes

Suburb             object
Address            object
Rooms               int64
Type               object
Price             float64
Method             object
SellerG            object
Date               object
Distance          float64
Postcode          float64
Bedroom2          float64
Bathroom         category
Car              category
Landsize          float64
Lattitude         float64
Longtitude        float64
Regionname         object
Propertycount     float64
dtype: object

In [112]:
df['Car'].unique()

[1.0, 0.0, 2.0, 6.0, 5.0, ..., 8.0, 7.0, 9.0, 10.0, NaN]
Length: 12
Categories (11, float64): [1.0, 0.0, 2.0, 6.0, ..., 8.0, 7.0, 9.0, 10.0]

In [114]:
df['Car'].mode()

0    2.0
Name: Car, dtype: category
Categories (11, float64): [0.0, 1.0, 2.0, 3.0, ..., 7.0, 8.0, 9.0, 10.0]

In [113]:
df['Car'].value_counts()

2.0     5606
1.0     5515
0.0     1026
3.0      748
4.0      507
5.0       63
6.0       54
8.0        9
7.0        8
10.0       3
9.0        1
Name: Car, dtype: int64

In [115]:
df['Car'].fillna(2.0, inplace = True)

In [117]:
df['Bathroom'].value_counts()

1.0    7517
2.0    4987
3.0     921
4.0     106
0.0      34
5.0      28
6.0       5
8.0       2
7.0       2
Name: Bathroom, dtype: int64

In [118]:
df['Bathroom'].fillna(1.0, inplace=True)

In [119]:
df.loc[:, ['Bathroom', 'Car']].describe()

Bathroom      Car
count    13603.0  13603.0
unique       9.0     11.0
top          1.0      2.0
freq      7518.0   5669.0

In [120]:
df['Car'].isnull().any()

False

In [121]:
df['Bathroom'].isnull().any()

False